src/dmd/mtype.d

```diff
@@ -683,8 +683,17 @@ extern (C++) abstract class Type : RootObject
             }
             else if (t1n.ty == t2n.ty && t1n.implicitConvTo(t2n))
                 goto Lcovariant;
-            else if (t1n.ty == Tnull && t1n.implicitConvTo(t2n) && t1n.size() == t2n.size())
-                goto Lcovariant;
+            else if (t1n.ty == Tnull)
+            {
+                /* Runtime representation of `typeof(null)` is same as `void*`.
+                   While it may be implicitly convertible to dynamic arrays,
+                   associative arrays, and delegates.  A `void*` is not covariant
+                   with types whose runtime representation is a `struct { }`. */
```

Comment:
Could you be more explicit about what covariant conversions are not allowed? For example, the comment mentions a struct, but I don't see a struct in the added code. The code suggests that if it is a struct, it will execute `goto Lcovariant` which seems very much at odds with the comment. Nor do I see a struct in the added test cases.

Reply:
Delegates, Associative arrays, dynamic arrays... are all structs.

Reply:
De-Kenji-afied the comment a bit.